File: lib/phx_shadcn/components/radio_group.ex
defmodule PhxShadcn.Components.RadioGroup do
  @moduledoc """
  RadioGroup component mirroring shadcn/ui RadioGroup.

  Manages single selection across a set of radio buttons with keyboard navigation.
  Supports 3 state modes: client-only, hybrid, and server-controlled.

  Sub-components: `radio_group/1`, `radio_group_item/1`.

  ## State Modes

  - **Client-only** — no `value` or `on_value_change`. Pure JS, server unaware.
  - **Hybrid** — `on_value_change` set, no `value`. JS selects instantly + pushes event.
  - **Server-controlled** — `value` set. Server owns the state.

  ## Examples

      <.radio_group id="plan" default_value="comfort">
        <div class="flex items-center space-x-2">
          <.radio_group_item value="default" id="r1" />
          <.label for="r1">Default</.label>
        </div>
        <div class="flex items-center space-x-2">
          <.radio_group_item value="comfort" id="r2" />
          <.label for="r2">Comfortable</.label>
        </div>
        <div class="flex items-center space-x-2">
          <.radio_group_item value="compact" id="r3" />
          <.label for="r3">Compact</.label>
        </div>
      </.radio_group>
  """

  use Phoenix.Component
  import PhxShadcn.Cn

  # ── RadioGroup (root) ──────────────────────────────────────────────

  @group_base_classes [
    "grid gap-3"
  ]

  attr :id, :string, required: true
  attr :value, :string, default: nil
  attr :default_value, :string, default: nil
  attr :on_value_change, :any, default: nil
  attr :orientation, :string, default: "vertical", values: ~w(vertical horizontal)
  attr :disabled, :boolean, default: false
  attr :name, :string, default: nil
  attr :class, :any, default: []
  attr :rest, :global

  slot :inner_block, required: true

  def radio_group(assigns) do
    state_mode =
      cond do
        assigns.value != nil -> "server"
        assigns.on_value_change != nil -> "hybrid"
        true -> "client"
      end

    hidden_value =
      case {assigns.value, assigns.default_value} do
        {nil, nil} -> ""
        {nil, dv} -> dv || ""
        {v, _} -> v || ""
      end

    orientation_class =
      if assigns.orientation == "horizontal" do
        "flex flex-row gap-3"
      else
        nil
      end

    assigns =
      assigns
      |> assign(:state_mode, state_mode)
      |> assign(:hidden_value, hidden_value)
      |> assign(
        :computed_class,
        cn([@group_base_classes, orientation_class, assigns.class])
      )

    ~H"""
    <div
      id={@id}
      role="radiogroup"
      data-slot="radio-group"
      data-state-mode={@state_mode}
      data-value={@value}
      data-default-value={@default_value}
      data-on-value-change={@on_value_change}
      data-orientation={@orientation}
      data-disabled={@disabled && "true"}
      aria-orientation={@orientation}
      class={@computed_class}
      phx-hook="PhxShadcnRadioGroup"
      {@rest}
    >
      {render_slot(@inner_block)}
      <input :if={@name} type="hidden" name={@name} value={@hidden_value} />
    </div>
    """
  end

  # ── RadioGroupItem ─────────────────────────────────────────────────

  @item_base_classes [
    "border-input text-primary focus-visible:border-ring focus-visible:ring-ring/50",
    "aspect-square size-4 shrink-0 rounded-full border shadow-xs",
    "transition-[color,box-shadow] outline-none focus-visible:ring-[3px]",
    "disabled:cursor-not-allowed disabled:opacity-50"
  ]

  @indicator_classes "relative flex items-center justify-center"

  attr :value, :string, required: true
  attr :id, :string, default: nil
  attr :disabled, :boolean, default: false
  attr :class, :any, default: []
  attr :rest, :global

  def radio_group_item(assigns) do
    assigns =
      assigns
      |> assign(:computed_class, cn([@item_base_classes, assigns.class]))
      |> assign(:indicator_classes, @indicator_classes)

    ~H"""
    <button
      type="button"
      role="radio"
      id={@id}
      data-slot="radio-group-item"
      data-roving-item
      data-value={@value}
      data-state="unchecked"
      data-disabled={@disabled && "true"}
      aria-checked="false"
      tabindex="-1"
      disabled={@disabled}
      class={@computed_class}
      {@rest}
    >
      <span data-slot="radio-group-indicator" class={@indicator_classes}>
        <svg
          class="fill-primary size-2"
          viewBox="0 0 8 8"
          xmlns="http://www.w3.org/2000/svg"
          style="display: none;"
          data-radio-circle
        >
          <circle cx="4" cy="4" r="4" />
        </svg>
      </span>
    </button>
    """
  end
end

File: priv/static/hooks/radio-group.js
/**
 * RadioGroup hook — manages single selection across radio buttons with keyboard nav.
 *
 * Mounts on the group <div> element. Uses event delegation for item clicks.
 * Supports 3 state modes: client, hybrid, server.
 * Inlines roving focus keyboard nav (arrow keys, Home/End, wrapping).
 *
 * Config via data attributes on the group:
 *   data-state-mode="client|hybrid|server"
 *   data-value="val" (server mode: current selected value)
 *   data-default-value="val" (client/hybrid: initial selected value)
 *   data-on-value-change="event_name|JS_struct"
 *   data-orientation="vertical|horizontal"
 *
 * External JS API — inbound CustomEvents:
 *   phx-shadcn:select — { detail: { value: "x" } }
 *   phx-shadcn:set    — { detail: { value: "x" } } (alias for select)
 *
 * Outbound CustomEvents (bubbles: false):
 *   phx-shadcn:selected — { id, value, action: "select" }
 */
import { notifyServer, listenForCommands, syncFormInput } from "./event-bridge.js";

const ITEM_SEL = "[data-slot=radio-group-item]";

function getEnabledItems(el) {
  return [...el.querySelectorAll(ITEM_SEL)].filter(
    (item) => item.dataset.disabled !== "true" && !item.disabled
  );
}

const RadioGroup = {
  mounted() {
    this.stateMode = this.el.dataset.stateMode || "client";
    this.eventCallback = this.el.dataset.onValueChange || null;
    this.orientation = this.el.dataset.orientation || "vertical";

    // Parse initial selected value
    this.selectedValue = null;
    const initial =
      this.stateMode === "server"
        ? this.el.dataset.value
        : this.el.dataset.defaultValue;

    if (initial) {
      this.selectedValue = initial;
    }

    // Apply initial state (skip form sync — hidden input already has the right value)
    this._initializing = true;
    this._applyState();
    this._initializing = false;

    // Click delegation
    this._onClick = (e) => {
      const item = e.target.closest(ITEM_SEL);
      if (!item) return;
      if (item.dataset.disabled === "true" || item.disabled) return;

      e.preventDefault();
      this._select(item.dataset.value, "user");
    };
    this.el.addEventListener("click", this._onClick);

    // Keyboard nav (roving focus inlined)
    this._onKeydown = (e) => {
      const item = e.target.closest(ITEM_SEL);
      if (!item) return;

      const isVertical = this.orientation === "vertical";
      const prev = isVertical ? "ArrowUp" : "ArrowLeft";
      const next = isVertical ? "ArrowDown" : "ArrowRight";

      let target = null;

      if (e.key === prev || e.key === "ArrowLeft" || e.key === "ArrowUp") {
        e.preventDefault();
        target = this._adjacentItem(item, -1);
      } else if (
        e.key === next ||
        e.key === "ArrowRight" ||
        e.key === "ArrowDown"
      ) {
        e.preventDefault();
        target = this._adjacentItem(item, 1);
      } else if (e.key === "Home") {
        e.preventDefault();
        const items = getEnabledItems(this.el);
        target = items[0] || null;
      } else if (e.key === "End") {
        e.preventDefault();
        const items = getEnabledItems(this.el);
        target = items[items.length - 1] || null;
      } else if (e.key === " ") {
        // Space selects the focused item (standard radio behavior)
        e.preventDefault();
        this._select(item.dataset.value, "user");
        return;
      }

      if (target) {
        target.focus();
        // Radio groups select on focus by default (WAI-ARIA pattern)
        this._select(target.dataset.value, "user");
      }
    };
    this.el.addEventListener("keydown", this._onKeydown);

    // Server->client push_event commands
    listenForCommands(this, ({ command, value }) => {
      switch (command) {
        case "select":
        case "set":
          this._select(value, "server");
          break;
      }
    });

    // External JS API — inbound CustomEvents
    this._onCustomSelect = (e) => {
      const value = e.detail && e.detail.value;
      if (value) this._select(value, "external");
    };
    this._onCustomSet = (e) => {
      const value = e.detail && e.detail.value;
      if (value) this._select(value, "external");
    };
    this.el.addEventListener("phx-shadcn:select", this._onCustomSelect);
    this.el.addEventListener("phx-shadcn:set", this._onCustomSet);
  },

  updated() {
    this.stateMode = this.el.dataset.stateMode || "client";
    this.eventCallback = this.el.dataset.onValueChange || null;
    this.orientation = this.el.dataset.orientation || "vertical";

    if (this.stateMode === "server") {
      this.selectedValue = this.el.dataset.value || null;
      requestAnimationFrame(() => this._applyState());
    }
  },

  destroyed() {
    this.el.removeEventListener("click", this._onClick);
    this.el.removeEventListener("keydown", this._onKeydown);
    this.el.removeEventListener("phx-shadcn:select", this._onCustomSelect);
    this.el.removeEventListener("phx-shadcn:set", this._onCustomSet);
  },

  // --- Core state transitions ---

  _select(value, source) {
    if (this.selectedValue === value) return;

    if (this.stateMode === "server" && source !== "server") {
      this._notifyChange(value, source);
      return;
    }

    this.selectedValue = value;
    this._applyState();
    this._notifyChange(value, source);
    this._dispatchOutbound("phx-shadcn:selected", value);
  },

  // --- Roving focus helpers ---

  _adjacentItem(current, direction) {
    const items = getEnabledItems(this.el);
    const idx = items.indexOf(current);
    if (idx === -1) return items[0] || null;
    const next = (idx + direction + items.length) % items.length;
    return items[next];
  },

  // --- State application ---

  _applyState() {
    const items = this.el.querySelectorAll(ITEM_SEL);
    items.forEach((item) => {
      const value = item.dataset.value;
      const checked = value === this.selectedValue;
      item.dataset.state = checked ? "checked" : "unchecked";
      item.setAttribute("aria-checked", String(checked));
      // Roving tabindex: selected item gets 0, rest get -1
      item.setAttribute("tabindex", "-1");
      // Show/hide the circle indicator
      const circle = item.querySelector("[data-radio-circle]");
      if (circle) circle.style.display = checked ? "" : "none";
    });

    // Set tabindex=0 on the selected item, or first enabled item
    const selected = this.selectedValue
      ? this.el.querySelector(
          `${ITEM_SEL}[data-value="${CSS.escape(this.selectedValue)}"]`
        )
      : null;

    if (selected) {
      selected.setAttribute("tabindex", "0");
    } else {
      const first = getEnabledItems(this.el)[0];
      if (first) first.setAttribute("tabindex", "0");
    }

    // Sync hidden input for form integration (skip during mount)
    if (!this._initializing) {
      syncFormInput(this.el, this.selectedValue || "");
    }
  },

  // --- Notification ---

  _notifyChange(value, source) {
    if (source === "server") return;
    if (this.stateMode === "client") return;

    const payload = {
      id: this.el.id,
      value: value,
    };

    notifyServer(this, this.eventCallback, payload);
  },

  _dispatchOutbound(eventName, value) {
    this.el.dispatchEvent(
      new CustomEvent(eventName, {
        detail: { id: this.el.id, value: value, action: "select" },
        bubbles: false,
      })
    );
  },
};

export { RadioGroup as PhxShadcnRadioGroup };
export default RadioGroup;
